ejb/WFLY-17986.adoc

```diff
@@ -0,0 +1,234 @@
+= Support external EJB client access to EAP deployments on OpenShift over http
+:author:            Richard Achmatowicz
+:email:             rachmato@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+Existing EAP customers use the EJB client library to access EJB-based deployments (SFSBs, SLSBs, MDBs) on EAP
+on bare metal and such access needs to be provided whether the EAP deployment is on bare metal or on OpenShift.
+
+This analysis document aims to highlight the issues and concerns in providing EJB client access to EJB-based
+deployments on EAP in OpenShift.
+
+=== Client location and protocol
+Generally, an EJB client application may be internal or external to the OpenShift cluster, and may use any one of
+three transport protocols supported by the EJB client: remote, remote+http, or http (or their encrypted variants).
+
+This issue specifically focuses on the case where the EJB client application is:
+
+* external to the OpenShift cluster
+* the transport protocol is http (or its encrypted variant https)
+
+For the other location / protocol combinations, please see the related issues
+(https://issues.redhat.com/browse/EAP7-2063[EAP7-2063],  https://issues.redhat.com/browse/EAP7-2064[EAP7-2064],
+https://issues.redhat.com/browse/EAP7-2065[EAP7-2065])
+
+=== Access to cluster
+One issue of relevance here is how an external EJB client application gains entry to the OpenShift cluster.
+
+NOTE: We sometimes use the word `cluster` to refer to the OpenShift cluster in which the EAP application is
+deployed; at other times, to refer to the set of Pods which make up the EAP deployment when the
+HA (high availability) server profile is used. We will try to refer either to the OpenShift cluster or
+the EAP cluster to avoid ambiguity.
+
+OpenShift provides a limited set of entry points for external applications to gain access to the cluster. For binary
+protocols, NodePorts are used to gain access. For protocols based on HTTP, routes or ingress controllers
+are available. A route is a per-deployment load balancer which is accessible external to the cluster.
+An ingress controller is a load balancer external to the cluster which can be shared by multiple deployments.
+
```

Comment:
Reading https://access.redhat.com/documentation/en-us/red_hat_jboss_enterprise_application_platform/7.4/html/getting_started_with_jboss_eap_for_openshift_container_platform/eap-operator-for-automating-application-deployment-on-openshift_default#jakarta-enterprise-beans-remoting-on-openshift_default we can find details about "[headless service](https://access.redhat.com/documentation/en-us/openshift_container_platform/3.6/html-single/architecture/#headless-services)" created by the EAP Operator which, is not exposed outside of the cluster and, hence, not relevant for EAP7-2062;

On the other hand,  the EAP Operator creates a Route like e.g.:
```
kind: Route
apiVersion: route.openshift.io/v1
metadata:
  annotations:
    openshift.io/host.generated: 'true'
    wildfly.org/wildfly-server-generation: '1'
  resourceVersion: '27583543'
  name: eap-server-route
  uid: 4613b2f8-7ad6-4db7-8338-1109de1c1630
  creationTimestamp: '2023-07-13T10:24:49Z'
  managedFields:
    - manager: openshift-router
      operation: Update
      apiVersion: route.openshift.io/v1
      time: '2023-07-13T10:24:49Z'
      fieldsType: FieldsV1
      fieldsV1:
        'f:status':
          'f:ingress': {}
      subresource: status
    - manager: wildfly-operator
      operation: Update
      apiVersion: route.openshift.io/v1
      time: '2023-07-13T10:24:49Z'
      fieldsType: FieldsV1
      fieldsV1:
        'f:metadata':
          'f:annotations':
            .: {}
            'f:wildfly.org/wildfly-server-generation': {}
          'f:labels':
            .: {}
            'f:app.kubernetes.io/managed-by': {}
            'f:app.kubernetes.io/name': {}
            'f:app.openshift.io/runtime': {}
          'f:ownerReferences':
            .: {}
            'k:{"uid":"1d9d5b27-aa2b-4c9d-b267-cfdfc0c1ebf5"}': {}
        'f:spec':
          'f:port':
            .: {}
            'f:targetPort': {}
          'f:to':
            'f:kind': {}
            'f:name': {}
            'f:weight': {}
          'f:wildcardPolicy': {}
  namespace: eap7-2062
  ownerReferences:
    - apiVersion: wildfly.org/v1alpha1
      kind: WildFlyServer
      name: eap-server
      uid: 1d9d5b27-aa2b-4c9d-b267-cfdfc0c1ebf5
      controller: true
      blockOwnerDeletion: true
  labels:
    app.kubernetes.io/managed-by: eap-operator
    app.kubernetes.io/name: eap-server
    app.openshift.io/runtime: eap
spec:
  host: eap-server-route-eap7-2062.apps.tborgato-yhvd.eapqe.psi.redhat.com
  to:
    kind: Service
    name: eap-server-loadbalancer
    weight: 100
  port:
    targetPort: http
  wildcardPolicy: None
status:
  ingress:
    - host: eap-server-route-eap7-2062.apps.tborgato-yhvd.eapqe.psi.redhat.com
      routerName: default
      conditions:
        - type: Admitted
          status: 'True'
          lastTransitionTime: '2023-07-13T10:24:49Z'
      wildcardPolicy: None
      routerCanonicalHostname: router-default.apps.tborgato-yhvd.eapqe.psi.redhat.com
```
This Route is the only entry point to the EAP Cluster for clients outside the OpenShift cluster;

For this reason, unless we plan to ask for modifications to the EAP Operator, we should proceed relying on this route for EAP7-2062;

If you think this is correct, could you please incorporate this info into this AD?

Reply:
I light of our latest call, and the changes @rhusar is making to the Operator, we might want elaborate a little on this topic